### PR TITLE
specify SSE42 'target' attribute for Fast_CRC32()

### PR DESCRIPTION
**Pull request by tchaikov**
if we enable SSE42 globally when compiling the tree for preparing a
portable binary, which could be running on CPU w/o SSE42 instructions
even the GCC on the building host is able to emit SSE42 code, this leads
to illegal instruction errors on machines not supporting SSE42. to solve
this problem, crc32 detects the supported instruction at runtime, and
selects the supported CRC32 implementation according to the result of
`cpuid`. but intrinics like "_mm_crc32_u64()" will not be available
unless the "target" machine is appropriately specified in the command
line, like "-msse42", or using the "target" attribute.

we could pass "-msse42" only when compiling crc32c.cc, and allow the
compiler to generate the SSE42 instructions, but we are still at the
risk of executing illegal instructions on machines does not support
SSE42 if the compiler emits code that is not guarded by our runtime
detection. and we need to do the change in both Makefile and CMakefile.

or, we can use GCC's "target" attribute to enable the machine specific
instructions on certain function. in this way, we have finer grained
control of the used "target". and no need to change the makefiles. so
we don't need to duplicate the changes on both makefile and cmake as
the previous approach.

this problem surfaces when preparing a package for GNU/Linux distribution,
and we only applies to optimization for SSE42, so using a feature
only available on GCC/Clang is not that formidable.

### CODE REVIEW

File: util/crc32c.cc [hunk 1]
```diff
@@ -313,10 +313,19 @@ static inline void Slow_CRC32(uint64_t* l, uint8_t const **p) {
   table0_[c >> 24];
 }
 
+#if defined(HAVE_SSE42)
+#if defined(__GNUC__) && !defined(__clang__) && !defined(_WIN64)
```
Comment (tchaikov) [inline util/crc32c.cc]: the gcc-mingw on win32 does not offer FMV.

**Comment by tchaikov** [comment]
linking errors. will take a closer look.

**Comment by tchaikov** [comment]
it's a bug in binutils 2.24, see https://sourceware.org/bugzilla/show_bug.cgi?id=16467.

**Comment by tchaikov** [comment]
instead of using FMV which is broken due to https://sourceware.org/bugzilla/show_bug.cgi?id=16467, just use the plain runtime detection.

File: util/crc32c.cc [hunk 2]
```diff
@@ -375,10 +378,7 @@ static bool isSSE42() {
 #ifndef HAVE_SSE42
   return false;
 #elif defined(__GNUC__) && defined(__x86_64__) && !defined(IOS_CROSS_COMPILE)
-  uint32_t c_;
-  uint32_t d_;
-  __asm__("cpuid" : "=c"(c_), "=d"(d_) : "a"(1) : "ebx");
-  return c_ & (1U << 20);  // copied from CpuId.h in Folly.
+  return __builtin_cpu_supports("sse4.2");
```
[the remark this answers is not in the document]
Comment (tchaikov) [reply util/crc32c.cc]: thanks for your review. no, i cannot be sure about this. =/

clang supports this feature, see https://reviews.llvm.org/rL240994. but it was 2015. while clang has been supported C++11 since 3.3 back to 2013.

GCC 4.8+ supports this builtin function. hence GCC on mingw/cygwin does.   rocksdb uses C++11. and C++11 is well supported by GCC since 4.8. also it's a simple change only relies on GCC. so it's safe on GCC side.

in short, i will drop this change as it might break the SSE42 optimization on older clang versions.

**Comment by tchaikov** [comment]
@siying fixed and repushed.

File: util/crc32c.cc [hunk 3]
```diff
@@ -313,6 +313,9 @@ static inline void Slow_CRC32(uint64_t* l, uint8_t const **p) {
   table0_[c >> 24];
 }
 
+#if defined(HAVE_SSE42) && defined(__GNUC__)
+__attribute__ ((target ("sse4.2")))
```
Comment (tchaikov) [inline util/crc32c.cc]: see http://releases.llvm.org/3.8.0/tools/clang/docs/AttributeReference.html#target-gnu-target and https://gcc.gnu.org/onlinedocs/gcc-4.8.0/gcc/Function-Attributes.html

**Comment by tchaikov** [comment]
hey @siying mind taking a look again at your convenience? thanks.
